### PR TITLE
PWX-5747: consul failover retries

### PR DESCRIPTION
**Pull request by sdeoras**
Signed-off-by: Saurabh Deoras <sdeoras@gmail.com>

### CODE REVIEW

File: consul/kv_consul.go [hunk 1]
```diff
@@ -286,19 +368,30 @@ func (kv *consulKV) Create(
...
+				if ok && err == nil {
+					return kvPair, err
+				}
+				if _, err := kv.client.Session().Destroy(sessionPair.Session, nil); err != nil {
+					logrus.Error(err)
+				}
+				if _, err := kv.Delete(key); err != nil {
+					logrus.Error(err)
+				}
+				if err != nil {
```
Comment (adityadani) [inline consul/kv_consul.go]: To reduce code duplication we can add a new function retryOp, which takes in a callback function as an input which will invoke consul APIs respective to each API here.

File: consul/kv_consul.go [hunk 2]
```diff
@@ -324,17 +417,37 @@ func (kv *consulKV) Update(
...
 func (kv *consulKV) Enumerate(prefix string) (kvdb.KVPairs, error) {
 	prefix = kv.domain + prefix
 	prefix = stripConsecutiveForwardslash(prefix)
-	pairs, meta, err := kv.client.KV().List(prefix, nil)
+	var pairs api.KVPairs
+	var meta *api.QueryMeta
+	var err error
+	for range myParams.machines {
+		pairs, meta, err = kv.client.KV().List(prefix, nil)
+		// *** check error post loop
```
Comment (adityadani) [inline consul/kv_consul.go]: same here. We should check for specific errors before retrying on a different endpoint.

File: consul/kv_consul.go [hunk 3]
```diff
@@ -36,6 +36,19 @@ var (
...
+// param stores init paramaters for consul kv.
+type param struct {
+	domain       string
+	machines     []string
+	options      map[string]string
+	fatalErrorCb kvdb.FatalErrorCB
+}
+
+var (
+	// myParams is a global var used to access kv init params.
```
[the remark this answers is not in the document]
Comment (sdeoras) [reply consul/kv_consul.go]: done

File: consul/kv_consul.go [hunk 4]
```diff
@@ -139,6 +152,22 @@ func New(
...
+	myParams.domain = domain
+	myParams.machines = machines
+	myParams.options = options
+	myParams.fatalErrorCb = fatalErrorCb
+
+	return initKv(myParams)
+}
+
+func initKv(p param) (kvdb.Kvdb, error) {
+	domain := p.domain
```
[the remark this answers is not in the document]
Comment (sdeoras) [reply consul/kv_consul.go]: done

File: consul/kv_consul.go [hunk 5]
```diff
@@ -139,6 +152,22 @@ func New(
...
+	machines := p.machines
```
[the remark this answers is not in the document]
Comment (sdeoras) [reply consul/kv_consul.go]: done

File: consul/kv_consul.go [hunk 6]
```diff
@@ -181,10 +210,30 @@ func (kv *consulKV) Get(key string) (*kvdb.KVPair, error) {
...
 	key = stripConsecutiveForwardslash(key)
-	pair, meta, err := kv.client.KV().Get(key, options)
+
+	var pair *api.KVPair
+	var meta *api.QueryMeta
+	var err error
+	// try as many times as there are machines
+	for range myParams.machines {
+		pair, meta, err = kv.client.KV().Get(key, options)
+		if err != nil { // *** this error is checked below post loop
```
[the remark this answers is not in the document]
Comment (sdeoras) [reply consul/kv_consul.go]: yup, just added this logic. it is messy but take a look.

File: consul/kv_consul.go (hunk 1 above)
Comment (sdeoras) [reply consul/kv_consul.go]: perhaps for next pr. this pr was only to get failover working

File: consul/client.go [hunk 7]
```diff
@@ -0,0 +1,509 @@
...
+		connError, err = c.reconnectIfConnectionError(conn, err)
+		if connError {
+			continue
+		} else {
+			break
+		}
+	}
+
+	if !ok {
+		return nil, ok, fmt.Errorf("Failed to set ttl")
```
Comment (adityadani) [inline consul/client.go]: nit: we can return the last seen error in the above loop here - fmt.Errorf("Failed to set ttl: %v", err)

File: consul/client.go [hunk 8]
```diff
@@ -0,0 +1,509 @@
...
+	var err error
+	connError := false
+
+	for i := 0; i < c.maxRetries; i++ {
+		conn := c.conn
+		ok, meta, err = conn.client.KV().Acquire(p, q)
+		if ok && err == nil {
+			return nil, ok, err
+		}
+		if _, err := conn.client.Session().Destroy(p.Session, nil); err != nil {
```
Comment (adityadani) [inline consul/client.go]: why are we doing the destroy and delete operations with the old connection before reconnecting?

File: consul/kv_consul.go [hunk 9]
```diff
@@ -424,12 +408,18 @@ func (kv *consulKV) CompareAndSet(
...
 	}
 
-	ok, _, err := kv.client.KV().CAS(pair, nil)
+	ok, _, err := kv.client.CompareAndSet(kvp.Key, kvp.Value, pair, nil)
 	if err != nil {
 		return nil, err
 	}
 
 	if !ok {
+		kvp, getErr := kv.Get(pair.Key)
```
Comment (adityadani) [inline consul/kv_consul.go]: we already do a similar Get and byte compare in the consulClient CompareAndSet implementation. We can avoid one extra Get operation by returning the kvp value in its implementation.

File: consul/kv_consul.go [hunk 10]
```diff
@@ -930,7 +919,7 @@ func (kv *consulKV) watchTreeStart(
 	checkIndex := func(prevIndex *uint64, pair *api.KVPair, newIndex uint64,
 		msg string, lastIndex, waitIndex uint64) {
 		if *prevIndex != 0 && newIndex <= *prevIndex {
-			kv.FatalCb(msg+" with index invoked twice: %v, prevIndex: %d"+
+			logrus.Infof(msg+" with index invoked twice: %v, prevIndex: %d"+
```
Comment (adityadani) [inline consul/kv_consul.go]: any reason why changing it from a panic callback to log?

Comment (sdeoras) [reply consul/kv_consul.go]: invalid connection was triggering fatal cb, which is now being mitigated by connection retry.

File: consul/client.go [hunk 11]
```diff
@@ -0,0 +1,509 @@
...
+	// httpError is a substring returned by consul during such http errors.
+	// Ideally such errors should be provided as consul constants
+	httpError = "Unexpected response code: 500"
+	// eofError is also a substring returned by consul during EOF errors.
+	eofError = "EOF"
+	// connRefused connection refused
+	connRefused = "getsockopt: connection refused"
+)
+
+const (
```
[the remark this answers is not in the document]
Comment (sdeoras) [reply consul/client.go]: done

File: consul/client.go (hunk 7 above)
Comment (sdeoras) [reply consul/client.go]: done

**Pull request merged**